### PR TITLE
GeometryInfo: Abstraction of instrument/geometry for simple parameters (L1, L2, 2theta, ...)

### LINKED ISSUE #15294: Basic instrument / geometry abstraction for use in algorithms.
Create a simple abstraction that allows getting things like L1, L2, phi in algorithms. This will remove the need for including instrument headers in algorithms, will be less repetitive, and easier to test.

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Closes #15294.

This pull request contains the basic class `GeometryInfoFactory` and `GeometryInfo`, as well as a modified algorithm that use them, to demonstrate that this is actually useful.

The introduction of `GeometryInfoFactory` and `GeometryInfo` implies that we now have a central and unit-tested place for computing simple instrument geometry parameters. Currently code for computing quantities such as L1, L2, and 2theta is scattered across dozens of algorithms and rarely unit-tested.

No release notes, as this is an internal change (for now, see also #15357).

No specific testing, just code review. Note that we will implicitly be testing this via many of the algorithm unit tests, once this class is in wider use.

Note that these classes are not necessarily complete and might be extended as needed.

### CODE REVIEW

File: Framework/API/inc/MantidAPI/GeometryInfoFactory.h [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+class MANTID_API_DLL GeometryInfoFactory {
```
Comment (OwenArnold) [inline Framework/API/inc/MantidAPI/GeometryInfoFactory.h]: `GeometryInfoFactory` makes me think the sole responsibility of this is creational, i.e. is to create `GeometryInfo`, I think the relationship is actually more complex than this.

File: Framework/API/src/GeometryInfo.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  // latter, once GeometryInfo is in widespread use.
+  const Kernel::V3D samplePos = m_factory.getSamplePos();
+  const Kernel::V3D beamLine = samplePos - m_factory.getSourcePos();
+
+  if (beamLine.nullVector()) {
+    throw Kernel::Exception::InstrumentDefinitionError(
+        "Source and sample are at same position!");
+  }
+
+  return m_detector->getTwoTheta(samplePos, beamLine);
```
Comment (OwenArnold) [inline Framework/API/src/GeometryInfo.cpp]: The calls are still going via the detector objects. Does the GeometryInfo provide a faster lookup, or is it simply a facade on the current queries we perform?

File: Framework/TestHelpers/src/InstrumentCreationHelper.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
+#include "MantidTestHelpers/InstrumentCreationHelper.h"
```
Comment (OwenArnold) [inline Framework/TestHelpers/src/InstrumentCreationHelper.cpp]: Now that you've created this missing helper. I think the simple instruments such as [this](https://github.com/mantidproject/mantid/blob/15294_geometry_info/Framework/TestHelpers/src/WorkspaceCreationHelper.cpp#L460-L485) can also eventually live there.

**Comment by OwenArnold** [comment]
My idea was to try to create and maintain an l1, l2 cache inside the Instrument. So instead of doing this:

```
auto det = outputWS->getDetector(workspaceIndex);
l2 = det->getDistance(*sample);
```
This could be done directly on the instrument:
```
auto l2 = instrument->l2(detectorId)
```
The workspaceIndex -> detectorId mapping would still have to be stored on the Workspace somehow, but the `instrument->l2(detectorId) would be a direct read of a cached array entry.

I don't know how/if this would affect what's been done here, what the performance would be like, and which approach is better. Shall I continue to explore my approach or did you already consider that?

File: Framework/API/src/GeometryInfo.cpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+bool GeometryInfo::isMonitor() const { return m_detector->isMonitor(); }
+
+bool GeometryInfo::isMasked() const { return m_detector->isMasked(); }
+
+double GeometryInfo::getL1() const { return m_factory.getL1(); }
+
+double GeometryInfo::getL2() const {
+  if (!isMonitor()) {
+    auto &sample = m_factory.getSample();
+    return m_detector->getDistance(sample);
```
Comment (OwenArnold) [inline Framework/API/src/GeometryInfo.cpp]: For the correct representation of the neutronic instrument. This is definitely something we want to shield from the client when they request l2(). Complex beam paths will have to be considered in future geometries, which will require access to the full instrument to "evaluate" the exact neutronic path across the scattering components.

**Comment by OwenArnold** [comment]
I think it would be wise to add/update performance tests for this new stuff.

File: Framework/API/inc/MantidAPI/GeometryInfo.h [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+class MANTID_API_DLL GeometryInfo {
+public:
+  /// Constructor, usually not used directy. Creation of GeometryInfo is done
+  /// via GeometryInfoFactory.
+  GeometryInfo(const GeometryInfoFactory &factory, const ISpectrum &spectrum);
+
+  /// Returns true is the spectrum is a monitor.
+  bool isMonitor() const;
```
Comment (OwenArnold) [inline Framework/API/inc/MantidAPI/GeometryInfo.h]: I do like the `isMonitor` and `isMasked` abstractions.

File: Framework/API/src/GeometryInfo.cpp (hunk 2 above)
Comment (SimonHeybrock) [inline Framework/API/src/GeometryInfo.cpp]: Yes, maybe that was a misunderstanding: It is simply a facade and was targeting at removing duplicated code from algorithms. The only thing that might make it marginally faster is that I am getting the instrument only once in the factory, and not as previously in each `MatrixWorkspace::getDetector` call.

File: Framework/TestHelpers/src/InstrumentCreationHelper.cpp (hunk 3 above)
Comment (SimonHeybrock) [inline Framework/TestHelpers/src/InstrumentCreationHelper.cpp]: :+1:

File: Framework/API/src/GeometryInfo.cpp (hunk 4 above)
Comment (SimonHeybrock) [inline Framework/API/src/GeometryInfo.cpp]: Quite a few algorithms also have flags `EMode` and `EFixed` and then read certain values from the parameter map, I guess that is also related?

**Comment by SimonHeybrock** [comment]
@OwenArnold:
"I don't know how/if this would affect what's been done here, what the performance would be like, and which approach is better. Shall I continue to explore my approach or did you already consider that?"

I think you should certainly go on exploring. What I did here had a different goal (reduce code duplication in algorithms). My hope was that this would later make the instrument rewrite easier to do, by having fewer algorithm that depend directly on it.

Do you think there is a simple and useful way to integrate this more with the instrument? If not, is there a way I should adapt this, such that it will make the next step (the full instrument rewrite) easier?

**Comment by OwenArnold** [comment]
@SimonHeybrock What I'm thinking at the moment is that the `GeometryInfo` is actually a really nice thing to return from the `MatrixWorkspace` and the `MatrixWorkspace` already has too much responsibility. I'm going to try to explore a slightly different approach, but I think we'll be able to converge in the end (a I don't know that my approach will be better/worse yet).

**Comment by OwenArnold** [comment]
:shipit:

**Comment by martyngigg** [comment]
:+1:

**Pull request merged**